### ISSUE TITLE
Vaski: Maksujen päivämäärien korjaus järjestelmänvaihdon aikaisille maksuille

### ISSUE THREAD

**Issue by MikkoLiimatainen** [KohaSuomi/Koha#1003]
### Tukipyyntö

Huomasimme, että järjestelmänvaihdon ajalta on osa maksuista kirjautunut päivämäärälle 4.6.2021, vaikka maksut ovat voineet olla paljonkin vanhempia. Monet näistä maksuista ovat jo vanhentuneita, mutta eivät ole poistuneet ajossa, koska päivämäärä on väärin.

Hain maksut, jotka pitäisi korjata, alla olevalla raportilla. Laskin maksun kuvauksessa olevan eräpäivän perusteella maksuille todellisen määräytymispäivän, joka haluttaisiin maksujen date-kenttään.

`SELECT STR_TO_DATE(CONCAT(LEFT(RIGHT(description, 16),10), ' 00:00:00'), '%d.%m.%Y %H:%i:%s') + INTERVAL 30 DAY AS maksunpäivä,
	al.*
FROM accountlines al
WHERE debit_type_code IN ('OVERDUE')
    AND DATE(date) = '2021-06-04'
    AND amountoutstanding > 0`

Onnistuuko tällainen korjaus ja voiko maksuille sen jälkeen ajaa vanhenemisajon, jotta saataisiin näistä jo vanhentuneet maksut poistettua?

**Comment by lmstrand** [comment]
Ajettu:
```
MariaDB [vaskiprod]> update accountlines set date = ( SELECT STR_TO_DATE(CONCAT(LEFT(RIGHT(description, 16),10), ' 00:00:00'), '%d.%m.%Y %H:%i:%s') + INTERVAL 30 DAY ) WHERE debit_type_code IN ('OVERDUE') AND DATE(date) = '2021-06-04' AND amountoutstanding > 0;

Query OK, 4297 rows affected (4.614 sec)
Rows matched: 4301  Changed: 4297  Warnings: 0
```

**Comment by lmstrand** [comment]
Ajastan 5v maksujen mitätöintiajon ajautumaan täksi yöksi.

**Comment by MikkoLiimatainen** [comment]
Tämä asia näyttää nyt olevan kunnossa. Kiitos.

**Comment by MikkoLiimatainen** [comment]
Olisikohan tämä mahdollista tehdä uudelleen? Nuo epäonnistuneet mitätöinnit kun loivat uudet maksurivit, niin minusta niille voisi laittaa taas oikeat päivämäärät ja kokeilla sitten korjatun vanhentamisskriptin toimintaa.

Tällä raportilla saa muutettavat rivit ja niille oikeat maksunmuodostuspäivä.
`SELECT REGEXP_SUBSTR(description, '[0-9]{2}.[0-9]{2}.[0-9]{4}') AS eräpäivä,
    STR_TO_DATE(CONCAT(REGEXP_SUBSTR(description, '[0-9]{2}.[0-9]{2}.[0-9]{4}'), ' 00:00:00'), '%d.%m.%Y %H:%i:%s') + INTERVAL 30 DAY AS maksunpäivä,
    al.*
FROM accountlines al
WHERE borrowernumber IN (SELECT borrowernumber FROM issues WHERE date_due + INTERVAL 30 DAY < '2019-01-01' AND itemnumber IN (SELECT itemnumber FROM items WHERE itype NOT IN ('14VRKL','14VRKLEHL','14VRKPELIL','28VRKAVL','28VRKL','42VRK')))
	AND timestamp between '2024-01-12 00:00:00' and '2024-01-12 12:00:00' and status = 'UNRETURNED' and debit_type_code = 'OVERDUE' and interface = 'cron' and amountoutstanding = '9.000000'`

**Comment by lmstrand** [comment]
ajettu:

```
MariaDB [vaskiprod]> update accountlines set date = STR_TO_DATE(CONCAT(REGEXP_SUBSTR(description, '[0-9]{2}.[0-9]{2}.[0-9]{4}'), ' 00:00:00'), '%d.%m.%Y %H:%i:%s') + INTERVAL 30 DAY
    -> where accountlines_id in
    -> (select accountlines_id FROM accountlines al WHERE borrowernumber IN (SELECT borrowernumber FROM issues WHERE date_due + INTERVAL 30 DAY < '2019-01-01' AND itemnumber IN (SELECT itemnumber FROM items WHERE itype NOT IN ('14VRKL','14VRKLEHL','14VRKPELIL','28VRKAVL','28VRKL','42VRK'))) AND timestamp between '2024-01-12 00:00:00' and '2024-01-12 12:00:00' and status = 'UNRETURNED' and debit_type_code = 'OVERDUE' and interface = 'cron' and amountoutstanding = '9.000000'
    -> );
Query OK, 2063 rows affected (0.314 sec)
Rows matched: 2063  Changed: 2063  Warnings: 0
```
2067 muuttui oikeasti, testasin ensin yhdellä borrowernumberilla.

**Comment by lmstrand** [comment]
Ja nyt on ajettu uusiksi 5v maksujen mitätöinti korjatulla skriptillä.

**Comment by MikkoLiimatainen** [comment]
No, nyt en löydä enää yhtään uudelleen muodostunutta maksua tai maksuja jotka olisivat jääneet vanhentumatta. Tämä näyttää siis olevan kunnossa.
